### PR TITLE
Fix forced alignment on patch version

### PR DESCRIPTION
**Pull request by melix**
### Context

See https://github.com/nebula-plugins/gradle-nebula-integration/issues/51

### CODE REVIEW

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/VirtualPlatformState.java [hunk 1]
```diff
@@ -57,30 +55,46 @@ void participatingModule(ModuleResolveState state) {
...
+
     List<String> getCandidateVersions() {
+        String forcedVersion = getForcedVersion();
         ComponentState selectedPlatformComponent = platformModule.getSelected();
-        if (selectedPlatformComponent.getSelectionReason().isForced()) {
-            return Collections.singletonList(selectedPlatformComponent.getVersion());
-        }
-        List<String> sorted = Lists.newArrayListWithCapacity(participatingModules.size());
+        List<String> sorted = Lists.newArrayListWithCapacity(participatingModules.size() + 1);
+        sorted.add(selectedPlatformComponent.getVersion());
```
Comment (ljacomet) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/VirtualPlatformState.java]: ❓ I wonder if the `forcedVersion` computed above does not need to be added as well when it is not `null` and different than the `selectedPlatformComponent.getVersion()`.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyGraphBuilder.java [hunk 2]
```diff
@@ -325,26 +327,45 @@ private void validateGraph(ResolveState resolveState) {
...
+        List<EdgeState> forcedEdges = null;
+        boolean hasMultipleVersions = false;
+        String currentVersion = module.maybeFindForcedPlatformVersion();
         Set<ModuleResolveState> participatingModules = module.getPlatformState().getParticipatingModules();
         for (ModuleResolveState participatingModule : participatingModules) {
             for (EdgeState incomingEdge : participatingModule.getIncomingEdges()) {
-                if (incomingEdge.getSelector().isForce()) {
+                SelectorState selector = incomingEdge.getSelector();
+                if (selector.isForce() && !selector.isSoftForce()) {
                     // Filter out platform originating edges
```
Comment (ljacomet) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyGraphBuilder.java]: That comment should either be removed or move down above line 349

Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyGraphBuilder.java]: Even better, the condition could be extracted to a method named `isPlatformOriginatingEdge` and the comment removed.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/selectors/SelectorStateResolver.java [hunk 3]
```diff
@@ -67,6 +68,14 @@ public T selectBest(ModuleIdentifier moduleId, List<? extends ResolvableSelector
             return candidates.get(0);
         }
 
+        List<T> allowed = candidates.stream().filter((v) -> SelectorStateResolverResults.isAllowedVersion(v)).collect(Collectors.toList());
```
Comment (ljacomet) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/selectors/SelectorStateResolver.java]: ❓ Method reference for that `filter` step

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/VirtualPlatformState.java (hunk 1 above)
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/VirtualPlatformState.java]: I wonder too, but I didn't find any case where it does...

**Comment by melix** [comment]
@ljacomet don't trust the :x: above, performance tests passed with the latest commit. Let me know what you think.

**Comment by ljacomet** [comment]
@melix Can you add the release note reference to the contributed test?

**Comment by melix** [comment]
@ljacomet he already is ;)

**Pull request merged**